### PR TITLE
Misc work

### PR DESCRIPTION
**Pull request by est77**
Added BaseGroup and Assembly clear method: 
- Useful to implement interactive updates of instancers in plugins.

Embree cleanups:
- Remove extra ray origin and dir copy.
- Remove unnecesary impl classes and levels of indirection to access some member variables.
- Move APPLESEED_WITH_EMBREE checks where they are needed, like other optional features.

### CODE REVIEW

File: src/appleseed/renderer/kernel/intersection/assemblytree.h [hunk 1]
```diff
@@ -32,7 +32,9 @@
 
 // appleseed.renderer headers.
 #include "renderer/kernel/intersection/curvetree.h"
+#ifdef APPLESEED_WITH_EMBREE
```
Comment (fedormatantsev) [inline src/appleseed/renderer/kernel/intersection/assemblytree.h]: I don't know if it's necessary here. The intent was to hide this check inside the header to keep outer code cleaner.

File: src/appleseed/renderer/kernel/intersection/embreescene.cpp [hunk 2]
```diff
@@ -250,17 +221,20 @@ namespace
         const ShadingRay&   shading_ray,
         RTCRay&             embree_ray)
     {
-        const RayBuffer buffer(shading_ray.m_org, shading_ray.m_dir);
+        embree_ray.org_x = static_cast<float>(shading_ray.m_org.x);
```
Comment (fedormatantsev) [inline src/appleseed/renderer/kernel/intersection/embreescene.cpp]: Can you explain this revert?

File: src/appleseed/renderer/kernel/intersection/embreescene.h [hunk 3]
```diff
@@ -42,6 +41,9 @@
 #include "foundation/utility/poolallocator.h"
 #include "foundation/utility/uid.h"
 
+// Embree headers.
+#include <embree3/rtcore.h>
```
Comment (fedormatantsev) [inline src/appleseed/renderer/kernel/intersection/embreescene.h]: If this move is made, no need in EmbreeDevice object at all. It was introduce to hide embree guts from outer code.

File: src/appleseed/renderer/kernel/intersection/intersector.h [hunk 4]
```diff
@@ -32,7 +32,9 @@
 
 // appleseed.renderer headers.
 #include "renderer/kernel/intersection/curvetree.h"
+#ifdef APPLESEED_WITH_EMBREE
 #include "renderer/kernel/intersection/embreescene.h"
```
Comment (fedormatantsev) [inline src/appleseed/renderer/kernel/intersection/intersector.h]: Again, I vote for no checks in outer code, check is done inside EmbreeScene.h

File: src/appleseed/renderer/kernel/intersection/assemblytree.h (hunk 1 above)
Comment (est77) [reply src/appleseed/renderer/kernel/intersection/assemblytree.h]: We do the check outside for all our others optional dependencies.
I did this change for consistency.

File: src/appleseed/renderer/kernel/intersection/embreescene.cpp (hunk 2 above)
Comment (est77) [reply src/appleseed/renderer/kernel/intersection/embreescene.cpp]: You were casting and copying manually the org and dir from shading_ray to RayBuffer only to copy them again to the embree ray, this time using mm_store_ps.

You can copy them directly to the embree ray and avoid the store_ps intrinsics.

File: src/appleseed/renderer/kernel/intersection/embreescene.h (hunk 3 above)
Comment (est77) [reply src/appleseed/renderer/kernel/intersection/embreescene.h]: EmbreeScene is not part of the public API and it is not included by any public header.
Removing the private implementation saves a memory allocation and one indirection when accessing the class members.

I still think that EmbreeDevice can be useful as a wrapper, if we want to set options in the embree device, so I didn't remove it.

File: src/appleseed/renderer/kernel/intersection/intersector.h (hunk 4 above)
Comment (est77) [reply src/appleseed/renderer/kernel/intersection/intersector.h]: I prefer them outside myself, it is more readable for me. But it is not a big preference.
What I don't like is that we check outside for some optional dependencies and inside for embree.

File: src/appleseed/renderer/kernel/intersection/embreescene.cpp (hunk 2 above)
Comment (fedormatantsev) [reply src/appleseed/renderer/kernel/intersection/embreescene.cpp]: Ah, okay. This was done based on embree's performance tips. They are in documentation pdf located somewhere in embree distribution folder in built appleseed deps

File: src/appleseed/renderer/kernel/intersection/embreescene.h (hunk 3 above)
Comment (fedormatantsev) [reply src/appleseed/renderer/kernel/intersection/embreescene.h]: Sounds reasonable, nice!

File: src/appleseed/renderer/kernel/intersection/intersector.h (hunk 4 above)
Comment (fedormatantsev) [reply src/appleseed/renderer/kernel/intersection/intersector.h]: Okay, then you're safe to remove this check inside the header itself

File: src/appleseed/renderer/modeling/scene/basegroup.cpp [hunk 5]
```diff
@@ -98,6 +98,16 @@ ShaderGroupContainer& BaseGroup::shader_groups() const
     return impl->m_shader_groups;
 }
 
+void BaseGroup::clear()
+{
+    impl->m_colors.clear();
+    impl->m_textures.clear();
+    impl->m_texture_instances.clear();
+    impl->m_shader_groups.clear();
+    impl->m_assemblies.clear();
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/scene/basegroup.cpp]: Do we acknowledge here that we're not calling `clear()` on sub-assemblies?

File: src/appleseed/renderer/kernel/intersection/embreescene.h [hunk 6]
```diff
@@ -96,6 +97,9 @@ class EmbreeScene
   private:
     struct Impl;
     Impl* impl;
+
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/intersection/embreescene.h]: Is the private implementation still useful/reasonable at this point?

File: src/appleseed/renderer/modeling/scene/basegroup.cpp (hunk 5 above)
Comment (est77) [reply src/appleseed/renderer/modeling/scene/basegroup.cpp]: We are simply deleting sub-assemblies.

File: src/appleseed/renderer/kernel/intersection/embreescene.h (hunk 6 above)
Comment (est77) [reply src/appleseed/renderer/kernel/intersection/embreescene.h]: Removed

File: src/appleseed/renderer/kernel/intersection/embreescene.cpp [hunk 7]
```diff
@@ -53,86 +53,54 @@
...
-
 using namespace foundation;
 using namespace renderer;
 using namespace std;
 
 namespace renderer
 {
 
-namespace
+class EmbreeGeometryData
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/intersection/embreescene.cpp]: Should be inside an anonymous namespace.

Comment (dictoon) [reply src/appleseed/renderer/kernel/intersection/embreescene.cpp]: Ah no sorry, disregard.

**Pull request merged**